stdlib/public/core/Int128.swift.gyb

```diff
@@ -63,6 +113,85 @@ extension _${U}Int128: CustomDebugStringConvertible {
   }
 }
 
+extension _${U}Int128 {
+  public init?<S: StringProtocol>(_ text: S, radix: Int = 10) {
+    guard !text.isEmpty else { return nil }
+    guard 2...36 ~= radix else { return nil }
+    self.init()
+    if let x : _UInt128 = _UInt128.value(from: text, radix: radix) {
+      self = Self(x)
+    } else {
+      return nil
+    }
+  }
+  
+  public init?(_ description: String) {
+    self.init(description, radix:10)
+  }
+}
+
+extension _${U}Int128 {
+  /// This method breaks `string` into pieces to reduce overhead of the
+  /// multiplications and allow UInt64 to work in converting larger numbers.
+  fileprivate static func value<S: StringProtocol>(
+      from string: S, radix: Int = 10) -> Self? {
+    // Handles signs and leading zeros
+    let uradix = UInt64(radix)
+    var s = String(string)
+    % if signed:
+    var isNegative = false
+    if s.hasPrefix("-") { isNegative = true }
+    % else:
+    if s.hasPrefix("-") { return nil }
+    % end
+    if s.hasPrefix("+") { s.removeFirst() }
+    while s.hasPrefix("0") { s.removeFirst() }
+    
+    // Translate the string into a number
+    var r = zero
+    let _ = times(x: zero, timesRadix: uradix, toPower: 1) // initialize table
+    while !s.isEmpty {
+      // handle `chunk`-sized digits at a time
+      let chunk = s.prefix(powers.count)
+      let size = chunk.count; s.removeFirst(size)
+      if let uint = UInt64(chunk, radix: radix) {
+        if size != 0 {
+          r = times(x: r, timesRadix: uradix, toPower: size)
+        }
+        r += Self(high: 0, low: uint)
+      } else {
+        return nil
+      }
+    }
+    % if signed:
+    if isNegative {
+      return -r
+    }
+    % end
+    return r
+  }
+  
+  /// Computed powers of rⁿ up to UInt64.max where r is the radix
+  private static var powers = [UInt64]()
+  
+  /// Multiplies `x` by rⁿ where r is the `radix` and returns the product
+  private static func times<T:FixedWidthInteger>(
+    x: T, timesRadix radix: UInt64, toPower n: Int) -> T {
+    // calculate the powers of the radix and store in a table
+    if powers.isEmpty || powers.first! != radix {
```

Comment:
This has an obvious toctou race condition, as well as a more subtle race condition on targets without total store ordering. I don't think that it's actually possible to use a lazily computed table of powers like this without an explicit lock. Given what the expected distribution of radices looks like, simply covering power-of-two bases with a precomputed table for base 10 would be a much more attractive solution.

Reply:
Thanks for the feedback! Will do a precomputed table for base 10.

Reply:
@stephentyrone, latest update should fix this race condition issue.